File: go/program/ops/opVariableDefinition.go
package ops

import (
	"github.com/2manyvcos/jpl/go/definition"
	"github.com/2manyvcos/jpl/go/jpl"
)

type opVariableDefinition struct{}

// { name: string, pipe: [op] }
func (opVariableDefinition) OP(runtime jpl.JPLRuntime, input any, params definition.JPLInstructionParams, scope jpl.JPLRuntimeScope, next jpl.JPLScopedPiper) ([]any, jpl.JPLError) {
	return runtime.ExecuteInstructions(params.Pipe, []any{input}, scope, jpl.JPLScopedPiperFunc(func(output any, scope jpl.JPLRuntimeScope) ([]any, jpl.JPLError) {
		return next.Pipe(input, scope.Next(&jpl.JPLRuntimeScopeConfig{Vars: map[string]any{params.Name: output}}))
	}))
}

// { name: string, pipe: function }
func (opVariableDefinition) Map(runtime jpl.JPLRuntime, params jpl.JPLInstructionParams) (definition.JPLInstructionParams, jpl.JPLError) {
	return definition.JPLInstructionParams{
		Name: params.Name,
		Pipe: call(params.Pipe),
	}, nil
}

File: js/lib/program/ops/opVariableDefinition.ts
import type { JPLInstructionParams, JPLRuntimeScope } from '@/library';
import type { UserJPLInstructionParams } from '@/program/params';
import type JPLRuntime from '@/runtime';
import { call } from './utils';

export default {
  /** { name: string, pipe: [op] } */
  op(
    runtime: JPLRuntime,
    input: unknown,
    params: JPLInstructionParams,
    scope: JPLRuntimeScope,
    next: (
      output: unknown,
      scope: JPLRuntimeScope,
    ) => Promise<unknown[]> | unknown[],
  ): Promise<unknown[]> {
    return runtime.executeInstructions(
      params.pipe ?? [],
      [input],
      scope,
      (output) =>
        next(input, scope.next({ vars: { [params.name ?? '']: output } })),
    );
  },

  /** { name: string, pipe: function } */
  map(
    runtime: JPLRuntime,
    params: UserJPLInstructionParams,
  ): JPLInstructionParams {
    return {
      name: runtime.assertType(params.name, 'string'),
      pipe: call(params.pipe),
    };
  },
};
